### PR TITLE
Prevent spam registrations by leveraging MailMojo's subscription URL

### PR DESCRIPTION
**Pull request by stianpr**
Instead of using the subscription endpoint through AJAX, we now use a
regular form POST request to the MailMojo subscription endpoint. The user
will then be presented with the reCAPTCHA process which will prevent spam
registrations.

### CODE REVIEW

File: src/mailmojo-widget.php [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <?php
-/*  Copyright 2013  Eliksir AS  (email : post@e5r.no)
+/*  Copyright 2017  Eliksir AS  (email : post@e5r.no)
```
Comment (asteinlein) [inline src/mailmojo-widget.php]: Are we in 2017 already? ;) Anyhow, having year here is pretty pointless, just get rid of the year (and in other files) for a general copyright.

File: src/languages/mailmojo-nb_NO.po [hunk 2]
```diff
@@ -3,138 +3,130 @@ msgid ""
...
+"Language: nb_NO\n"
 "MIME-Version: 1.0\n"
 "Content-Type: text/plain; charset=UTF-8\n"
 "Content-Transfer-Encoding: 8bit\n"
+"X-Generator: Poedit 1.8.11\n"
 
-#: ../mailmojo-widget.php:28
+#: mailmojo-plugin.php:115
+msgid "MailMojo Settings"
+msgstr "MailMojo innstillinger"
```
Comment (fdanielsen) [inline src/languages/mailmojo-nb_NO.po]: Realize this is old, but in Norwegian I'd prefer "MailMojo-innstillinger".

File: src/languages/mailmojo-nb_NO.po [hunk 3]
```diff
@@ -3,138 +3,130 @@ msgid ""
...
+
+#: mailmojo-plugin.php:117
+msgid ""
+"Enter the username of the MailMojo account where the mailing list you want "
+"signups to are located. After saving the changes, go to the Widgets menu in "
+"the Appearance section to configure your widget."
+msgstr ""
+"Skriv inn brukernavnet til din MailMojo-konto. Etter du har lagret "
```
Comment (fdanielsen) [inline src/languages/mailmojo-nb_NO.po]: "Etter at du har lagret"

File: src/languages/mailmojo-nb_NO.po [hunk 4]
```diff
@@ -3,138 +3,130 @@ msgid ""
...
+"endringene går du til Widgeter menyen under Utseende seksjonen for å "
```
Comment (fdanielsen) [inline src/languages/mailmojo-nb_NO.po]: "Widgeter-menyen under Utseende-seksjonen"

File: src/languages/mailmojo-nb_NO.po [hunk 5]
```diff
@@ -3,138 +3,130 @@ msgid ""
...
+"konfiguerer widgeten."
```
Comment (fdanielsen) [inline src/languages/mailmojo-nb_NO.po]: Spelling mistake: "konfigurerer"

File: src/mailmojo-widget.php [hunk 6]
```diff
@@ -61,8 +49,7 @@ public function initFiles () {
 	 * @param array $instance
 	 */
 	public function form ($instance) {
-		$mmApi = $this->mmPlugin->getApi();
-		if ($mmApi === null) {
+		if (empty($this->mmPlugin->username)) {
 			global $blog_id;
 			$adminUrl = get_admin_url($blog_id);
 			$output = "<p>%s <a href=\"{$adminUrl}options-general.php?page={$this->mmPlugin->getBasename()}\">%s</a>.</p>";
```
Comment (fdanielsen) [inline src/mailmojo-widget.php]: Gah, GitHub.. Can't add comment on the lines below for some reason.. I noticed it in the translations files, and it's annoying that the text is split up into two gettext calls below. Seems like an good improvement to make while already updating the translations.

File: src/mailmojo-widget.php [hunk 7]
```diff
@@ -176,56 +154,60 @@ public function update ($newInstance, $oldInstance) {
...
+<p class="field">
+	<label for="mailmojo_{$this->number}_name">%s:</label>
+	<input class="text" type="text" id="mailmojo_{$this->number}_name" name="name">
+</p>
 HTML;
 			$incname = sprintf($incname, __('Name', 'mailmojo'));
 		}
 
-		// Checkboxes for tags
+		$tags = "";
```
Comment (fdanielsen) [inline src/mailmojo-widget.php]: Use single quotes here too, as you do for `$desc` below.

File: src/mailmojo-widget.php [hunk 8]
```diff
@@ -176,56 +154,60 @@ public function update ($newInstance, $oldInstance) {
...
-		// Description
+		$desc = '';
 		if (!empty($instance['desc'])) {
 			$desc = "<p>{$instance['desc']}</p>";
 		}
 
+		$tagsadditional = '';
+		if (!empty($instance['tags'])) {
+			$tagsadditional = <<<HTML
+<input type="hidden" name="tagsadditional" value="{$instance['tags']}">
```
Comment (fdanielsen) [inline src/mailmojo-widget.php]: Hm, so we output the same tags both in tags and tagsadditional?

File: src/mailmojo-widget.php [hunk 9]
```diff
@@ -176,56 +154,60 @@ public function update ($newInstance, $oldInstance) {
...
 {$before_widget}
 	{$before_title}{$instance['title']}{$after_title}
 	$desc
-	<form method="post" id="mailmojo_form_{$this->number}" class="mailmojo_form">
-		<div class="field">
-			<label for="mailmojo_email">%s:</label>
-			<input class="text" type="text" id="mailmojo_email" name="mailmojo_email">
-		</div>
+	<form action="https://{$this->mmPlugin->username}.mailmojo.no/{$instance['listid']}/s"
+			method="post" id="mailmojo_form_{$this->number}" class="mailmojo_form">
```
Comment (fdanielsen) [inline src/mailmojo-widget.php]: Suggest aligning `id` attribute with the fields, so `mailmojo_{$this->number}_form`.

File: src/mailmojo-widget.php [hunk 10]
```diff
@@ -176,56 +154,60 @@ public function update ($newInstance, $oldInstance) {
...
 		$output = <<<HTML
 {$before_widget}
 	{$before_title}{$instance['title']}{$after_title}
 	$desc
-	<form method="post" id="mailmojo_form_{$this->number}" class="mailmojo_form">
-		<div class="field">
-			<label for="mailmojo_email">%s:</label>
-			<input class="text" type="text" id="mailmojo_email" name="mailmojo_email">
-		</div>
+	<form action="https://{$this->mmPlugin->username}.mailmojo.no/{$instance['listid']}/s"
```
Comment (fdanielsen) [inline src/mailmojo-widget.php]: Suggest extracting URL to a helper method.

File: src/mailmojo-widget.php (hunk 8 above)
Comment (stianpr) [inline src/mailmojo-widget.php]: A good point. That is actually an oversight by me. We doesn't support tagsadditional yet.  I'll remove this.

File: src/languages/mailmojo-nb_NO.po [hunk 11]
```diff
@@ -3,138 +3,127 @@ msgid ""
...
+
+#: mailmojo-plugin.php:117
+msgid ""
+"Enter the username of the MailMojo account where the mailing list you want "
+"signups to are located. After saving the changes, go to the Widgets menu in "
+"the Appearance section to configure your widget."
+msgstr ""
+"Skriv inn brukernavnet til din MailMojo-konto. Etter at du har lagret "
+"endringene går du til Widgeter-menyen under Utseende-seksjonen for å "
+"konfigurerer widgeten."
```
Comment (fdanielsen) [inline src/languages/mailmojo-nb_NO.po]: Hehe, my bad.. Should be "konfigurere".

File: src/mailmojo-widget.php [hunk 12]
```diff
@@ -207,25 +184,32 @@ public function widget ($args, $instance) {
 {$before_widget}
 	{$before_title}{$instance['title']}{$after_title}
 	$desc
-	<form method="post" id="mailmojo_form_{$this->number}" class="mailmojo_form">
-		<div class="field">
-			<label for="mailmojo_email">%s:</label>
-			<input class="text" type="text" id="mailmojo_email" name="mailmojo_email">
-		</div>
+	<form method="post" action="{$this->getMailMojoUrl($instance['listid'])}"
```
Comment (fdanielsen) [inline src/mailmojo-widget.php]: Suggest the name `getSubscribeUrl`.

**Pull request merged**